### PR TITLE
feat: Add support scalar operation

### PR DESCRIPTION
**Pull request by lovenoble**

### CODE REVIEW

File: fhevm/tee_interpreter.go [hunk 1]
```diff
@@ -25,7 +25,7 @@ func doOp(
 ) ([]byte, error) {
 	logger := environment.GetLogger()
 
-	lp, rp, lhs, rhs, err := extract2Operands(op, environment, input, runSpan)
+	isScalar, lp, rp, lhs, rhs, err := extract2Operands(op, environment, input, runSpan)
```
Comment (amaury1093) [inline fhevm/tee_interpreter.go]: could you change extract2Operands signature to have:

```suggestion
	lp, rp, lhs, rhs, isScalar, err := extract2Operands(op, environment, input, runSpan)
```

File: fhevm/tee_interpreter.go [hunk 2]
```diff
@@ -50,7 +50,12 @@ func doOp(
 	// Note that we do arithmetic operations on uint64, then we convert the
 	// result back to the FheUintType.
 	l := big.NewInt(0).SetBytes(lp.Value).Uint64()
-	r := big.NewInt(0).SetBytes(rp.Value).Uint64()
+	var r uint64
+	if !isScalar {
+		r = big.NewInt(0).SetBytes(rp.(*tee.TeePlaintext).Value).Uint64()
+	} else {
+		r = big.NewInt(0).SetBytes(rp.([]byte)).Uint64()
+	}
```
Comment (amaury1093) [inline fhevm/tee_interpreter.go]: Does zama have any tests for scalar? We should add tests to tee scalar ops too

File: fhevm/tee_interpreter.go [hunk 3]
```diff
@@ -282,35 +309,59 @@ func extract1Operands(op string, environment EVMEnvironment, input []byte, runSp
 	return &cp, ct, nil
 }
 
-func extract2Operands(op string, environment EVMEnvironment, input []byte, runSpan trace.Span) (*tee.TeePlaintext, *tee.TeePlaintext, *verifiedCiphertext, *verifiedCiphertext, error) {
+func extract2Operands(op string, environment EVMEnvironment, input []byte, runSpan trace.Span) (bool, *tee.TeePlaintext, any, *verifiedCiphertext, any, error) {
```
Comment (amaury1093) [inline fhevm/tee_interpreter.go]: We lose a lot of semantic information by using `any`. What's the actual types we are interested in? `tee.Plaintext` and `*big.Int`, is that right?

Should we create an interface? A sum type?

File: fhevm/tee_interpreter.go (hunk 2 above)
Comment (lovenoble) [reply fhevm/tee_interpreter.go]: Zama don't have any tests for scalar.

File: fhevm/tee_interpreter.go (hunk 1 above)
Comment (lovenoble) [reply fhevm/tee_interpreter.go]: Okay

File: fhevm/tee_interpreter.go (hunk 3 above)
Comment (lovenoble) [reply fhevm/tee_interpreter.go]: Okay. I'll consider that part.

File: fhevm/tee_interpreter.go (hunk 2 above)
Comment (lovenoble) [reply fhevm/tee_interpreter.go]: @amaury1093 Do you wanna I add tests to tee scalar ops too?

Comment (amaury1093) [reply fhevm/tee_interpreter.go]: Yes

File: fhevm/tee_interpreter.go (hunk 3 above)
Comment (lovenoble) [reply fhevm/tee_interpreter.go]: @amaury1093 Could you clarify this part?

Comment (lovenoble) [reply fhevm/tee_interpreter.go]: I'm not sure 100% what you want.

Comment (amaury1093) [reply fhevm/tee_interpreter.go]: I'm also not 100% what's the best solution,but I'm pretty sure there's a better one than `any`. If you can think of something, feel free to propose. Or else I'll take a deeper look next week.

Comment (lovenoble) [reply fhevm/tee_interpreter.go]: Okay.

**Review by amaury1093 — APPROVED** [review]
utACK. Would feel more comfortable once this is tested on a contract with TEE.sol

**Comment by lovenoble** [comment]
In calucating gas, I haven't considered the cases of scalar operations. So I created a new PR https://github.com/Inco-fhevm/fhevm-go/pull/27 to fix it

**Pull request merged**